### PR TITLE
Use the same FT load flags in HB as in FT

### PR DESCRIPTION
**Pull request by NiLuJe**
Avoids mismatches & broke-ass advances

### CODE REVIEW

**Comment by NiLuJe** [comment]
That sounds completely sane, and is what raqm does, FWIW.

Hopefully leads to much more consistent results w/ HB + hinting.

File: crengine/src/lvfntman.cpp [hunk 1]
```diff
@@ -992,9 +992,17 @@ class LVFreeTypeFace : public LVFont
         if (FT_Err_Ok == error) {
             if (_hb_font)
                 hb_font_destroy(_hb_font);
-            _hb_font = hb_ft_font_create(_face, 0);
+            _hb_font = hb_ft_font_create(_face, NULL);
             if (!_hb_font)
                 error = FT_Err_Invalid_Argument;
+            // Use the same load flags as we do when using FT directly, to avoid mismatching advances & raster
```
Comment (poire-z) [inline crengine/src/lvfntman.cpp]: I guess the added part should be in a `else {...}` (if _hb_font ok) (not sure what happens next when _hb_font fails :)

Comment (NiLuJe) [reply crengine/src/lvfntman.cpp]: Indeed! I mistakenly thought the error path was a return, for, err, some reason? :D.

**Comment by poire-z** [comment]
And, as the flags may change widths and lines layout, and previously cached calculations may not match how words will be drawn after that, you'd better increment this:
https://github.com/koreader/crengine/blob/bb7b70523796533b26f73b6afd24567387a542a9/crengine/src/lvtinydom.cpp#L66-L67
(although just toggling any setting and back will re-render the document correctly when one feels it looks ugly)

File: crengine/src/lvfntman.cpp [hunk 2]
```diff
@@ -1071,9 +1082,20 @@ class LVFreeTypeFace : public LVFont
         if (FT_Err_Ok == error) {
             if (_hb_font)
                 hb_font_destroy(_hb_font);
-            _hb_font = hb_ft_font_create(_face, 0);
-            if (!_hb_font)
+            _hb_font = hb_ft_font_create(_face, NULL);
+            if (!_hb_font) {
                 error = FT_Err_Invalid_Argument;
+            } else {
+                // Use the same load flags as we do when using FT directly, to avoid mismatching advances & raster
```
Comment (Frenzie) [inline crengine/src/lvfntman.cpp]: Ah, excellent!

So by default it uses the hinting as specified by the font or something? 'cause I only see auto and no hint.

Comment (NiLuJe) [reply crengine/src/lvfntman.cpp]: It defaults to FT_LOAD_DEFAULTS | FT_LOAD_NO_HINTING (i.e., no hinting)... for apparently complex legacy reasons (the one release that flipped the defaults lasted all of 24h, because ithat massively broke firefox :D).

Comment (Frenzie) [reply crengine/src/lvfntman.cpp]: I thought all relevant hinting patents had expired?

Could you link me to the Firefox thing? I must be out of the loop on that one. I guess the Mozilla dev thingy blog doesn't advertise their own failures. :-D

Comment (NiLuJe) [reply crengine/src/lvfntman.cpp]: ```https://github.com/harfbuzz/harfbuzz/issues/143```

Comment (Frenzie) [reply crengine/src/lvfntman.cpp]: Oh, I thought you meant within the past few weeks. I may have a vague recollection of that. :-)

**Comment by poire-z** [comment]
But if you use `FT_LOAD_RENDER` (which must just loads metrics) instead of `FT_LOAD_DEFAULT` (which must load glyphs), you get back the 12s ! :)

**Comment by poire-z** [comment]
Oups, no, you don't :( (forgot to uncomment the `hb_ft_font_set_load_flags`...)

**Comment by poire-z** [comment]
Well, it's superslow whatever the flags, except when you use the flags that are the defaults in harfbuzz:  `flags = FT_LOAD_DEFAULT | FT_LOAD_NO_HINTING;`
https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-ft.cc#L85

There's also somes notes marked TODO at https://github.com/harfbuzz/harfbuzz/blob/22defe0965adddaa09eebc13df7fa6c64e2abba3/src/hb-ft.cc#L42-L62
```
 *   - FreeType works in 26.6 mode.  Clients can decide to use that mode, and everything
 *     would work fine.  However, we also abuse this API for performing in font-space,
 *     but don't pass the correct flags to FreeType.  We just abuse the no-hinting mode
 *     for that, such that no rounding etc happens.  As such, we don't set ppem, and
 *     pass NO_HINTING as load_flags.  Would be much better to use NO_SCALE, and scale
 *     ourselves, like we do in uniscribe, etc.
```

So, revert the flags parts? Or do you have some other ideas?

Also, if we were to keep that, there's more to do: they are currently just set from the _hintingMode value at font load time - _hintingMode may be changed later when hinting is changed, and the flags would currently not be updated - it would need the same kind of thing done in 
https://github.com/koreader/crengine/blob/c02f59701128cef0cbc34fefeed126ce36260e37/crengine/src/lvfntman.cpp#L936-L941
which is called for each loaded font when the global hinting mode is changed by:
https://github.com/koreader/crengine/blob/c02f59701128cef0cbc34fefeed126ce36260e37/crengine/src/lvfntman.cpp#L2352-L2364

**Comment by NiLuJe** [comment]
I'll take a look at it later tonight (but, yeah, I smell a commenting out coming ;D).

**Comment by NiLuJe** [comment]
Okay, so, clearly, it's not viable as-is.

But the other behavior seems broken, which explains the crazy spacing variations.

(f.g., with the matched flags, I get a result that's basically identical to FT, but with proper kerning, the spacing of non-kerned pairs is untouched).

**Comment by NiLuJe** [comment]
Okay, a few comparison shots.

FeeType, AutoHint
ft_auto

FreeType linked against HarfBuzz, AutoHint
_fthb_auto

**Comment by NiLuJe** [comment]
FreeType, Native hinting
ft_native

FreeType linked against HarfBuzz, Native hinting
_fthb_native

**Comment by NiLuJe** [comment]
HarfBuzz, AutoHint
hb_auto

Harfbuzz, Native hinting
hb_native

Spacing is FUBAR.

**Comment by NiLuJe** [comment]
HarfBuzz, matching flags, Native
hb_flags_native

HarfBuzz, matching flags, AutoHint
hb_flags_auto

General spacing matches the FT result, but with kerning pairs apparently better obeyed.

**Comment by NiLuJe** [comment]
Now, shit starts getting weird.

HarfBuzz, matching flags, autohinter, with a FT linked against HB.
_full_auto

That's my preferred result, right there! And, again, general spacing matches the FT linked w/ HB results.

And now, for the weird one:
HarfBuzz, matching flags, native hinting, with a FT linked against HB.

_full_native

This one's a mess, plain and simple.
Not quite sure why, possibly because we're not actually passing the right flags for the native hinter (which can be roundabout to explicitly request).

**Comment by NiLuJe** [comment]
@poire-z : I don't quite get how we actually end up with the native hinter in CRE... Is that the default case, and hoping FT was built w/ native hinting enabled? Because that might need an explicit NO_AUTOHINTER flag...

**Comment by poire-z** [comment]
These tests were in the original code, and I assume they are allright when I read this in freetype.h:
```
   *   By default, hinting is enabled and the font's native hinter (see
   *   @FT_FACE_FLAG_HINTER) is preferred over the auto-hinter.  You can
   *   disable hinting by setting @FT_LOAD_NO_HINTING or change the
   *   precedence by setting @FT_LOAD_FORCE_AUTOHINT.  You can also set
   *   @FT_LOAD_NO_AUTOHINT in case you don't want the auto-hinter to be
   *   used at all.
```

**Comment by poire-z** [comment]
I really don't have an eye for catching what's wrong :) so take my comments for what they are :|

`HarfBuzz, matching flags, AutoHint` looks not like the previous to me, so not as good (`me ans` `h er` `actu ally`) which is the kind of strange things I noticed when testing and concluded that HarfBuzz and Auto hint = bad (that was on the emulator, it's not as bad on my Kobo).

And in both your `HarfBuzz, matching flags, autohinter/native, with a FT linked against HB`. the `fr om` in `from a big list` is worse than in others :|

It mostly looks better on eInk than on our computer screens. The only bothering thing on eInk are the occassional fat and stuck `ss` as in a**ss**ociation.

**Comment by NiLuJe** [comment]
@poire-z : Yeah, except FT can be built with no native hinting support at all, and has a mind of its own an whether it actually always do what you tell it to (mostly, when dealing with tricky fonts).

But, yeah, generally, nothing means native, but I tend to prefer being very very explicit when I *want* native hinting for sure ;).

I'll see if that makes any difference in this case, though ;).

---

`HarfBuzz, matching flags, AutoHint` It's of course not as FT alone, but is is *much* closer than without the flags ;).

`HarfBuzz, matching flags, native, with a FT linked against HB` is borked, for some reason.

The autohint variant, on the other hand, looks pretty damn good (and, specifically, that from might not look all that great, but it is identical to FT's version ;)).

**Comment by NiLuJe** [comment]
Quick'n dirty visualisation w/ IM's compare tool.

FT to HB w/ Flags
ft_to_hb

FT to FT+HB
ft_to_fthb

FT to HB w/ flags w/ FT+HB
ft_to_full

**Comment by NiLuJe** [comment]
So, not a whole lot of changes between FT's and my preferred version with all the bells & whistle, which is... good? ;).

**Comment by NiLuJe** [comment]
Without the flags, it's of course a whole different story ;).

ft_to_hb

**Comment by NiLuJe** [comment]
Yeah, with an explicit ban on the autohinter, HB+FT+HB+Native looks as it should.

full_forcenative

And the diff is very similar to the AutoHinter diff :).

ft_to_full_native

**Comment by NiLuJe** [comment]
Added that bit to #231, because it can't hurt even without HB ;).

**Comment by NiLuJe** [comment]
@poire-z : Do you know if CRe loads the fallback font too in these instances? Because hinting a CJK font will of course takes a crazy amount of time, which might explain why we're seeing such a massive performance hit. (That's not to undercut the fact that it is going to be a large performance it, even in the best of circumstances. Doing this right in an efficient way is *hard*, as can be seen by the amount of glue code Blink & Gecko both need...).

**Comment by NiLuJe** [comment]
Yeah, so, as-needed, which means it'd have been potentially even more hilariously slow on CJK content ;p.

**Comment by poire-z** [comment]
For science, I tried using `hb_font_create` instead of `hb_ft_font_create` to avoid any freetype interaction:
```diff
@@ -46,6 +46,7 @@
 #if USE_HARFBUZZ==1
 #include <hb.h>
 #include <hb-ft.h>
+#include <hb-ot.h>
 #include "lvhashtable.h"
 #endif

@@ -1082,7 +1083,18 @@ public:
         if (FT_Err_Ok == error) {
             if (_hb_font)
                 hb_font_destroy(_hb_font);
-            _hb_font = hb_ft_font_create(_face, NULL);
+//            _hb_font = hb_ft_font_create(_face, NULL);
+hb_blob_t *blob = hb_blob_create_from_file(_fileName.c_str());
+printf ("Opened font file %s: %u bytes long\n", _fileName.c_str(), hb_blob_get_length(blob));
+hb_face_t *face = hb_face_create(blob, index); // index);
+hb_blob_destroy(blob);
+blob = NULL;
+unsigned int upem = hb_face_get_upem(face);
+_hb_font = hb_font_create(face);
+//hb_font_set_scale (_hb_font, upem, upem);  // not good
+hb_font_set_scale (_hb_font, size*64, size*64);
+hb_ft_font_set_funcs (_hb_font);
+
             if (!_hb_font) {
                 error = FT_Err_Invalid_Argument;
             } else {
@@ -1094,7 +1106,7 @@ public:
                 } else if (_hintingMode == HINTING_MODE_DISABLED) {
                     flags |= FT_LOAD_NO_AUTOHINT | FT_LOAD_NO_HINTING;
                 }
-                hb_ft_font_set_load_flags(_hb_font, flags);
+//                hb_ft_font_set_load_flags(_hb_font, flags);
             }
         }
 #endif

```
but not a good idea at first :)
image

With the probably more correct `hb_font_set_scale (_hb_font, size*64, size*64);`:
image

This then does not involve any freetype load flags on the HB side, and it renders in 13s.
It looks fine to my eyes. Curious about yours with your test document :)

**Comment by poire-z** [comment]
Other option: see what patching our harfbuzz with that would do
```
https://github.com/harfbuzz/harfbuzz/pull/1082
```

**Comment by poire-z** [comment]
^^ might need more proper arguments to the `hb_font_set_scale() `call. Dunno if the `size*64` is enough or right :) there are more complex stuff, for example in:
```
https://github.com/harfbuzz/harfbuzz/issues/559
```
And it's unclear to me what `hb_ft_font_set_funcs()` does ...

**Comment by NiLuJe** [comment]
Native:
pz_native

Auto:
pz_auto

**Comment by NiLuJe** [comment]
It's not honoring FT advances at all, which is basically what using the hb_ft things with the wrong flags doe,s as is shown by the comparison shots (perfect match in both cases).*

HB to PZ (Auto)
hb_to_pz_auto

HB To PZ (Native)
hb_to_pz_native

**Comment by NiLuJe** [comment]
Possibly a bit faster (I can still hear my CPU fan start to spin up ;p.) with the PR in, but it's apparently not perfect, so I'm not sure how viable that'd be.

**Comment by poire-z** [comment]
(OK :| about my hb_font_create try, thanks for testing :)
With the harfbuzz PR: Just _a bit_ faster ? So not worth it? But you got the results you liked?

**Comment by NiLuJe** [comment]
@poire-z : I don't really have hard numbers since it was only spinning the CPU for a few seconds with only Noto Serif & Noto Sans being loaded ;).

So, faster, yeah, but I can't really quantify it ;).

Results were identical as without the PR, AFAICT, but I'll double-check.

**Comment by NiLuJe** [comment]
Ah. If you look at the page, stuff renders indendtically, but there might be something wonky going on somewhere, because positioning/linge-spacing is going bonkers:

pr_auto

**Comment by NiLuJe** [comment]
Take that w/ a grain of salt, I moved that build to my base head, so that's HB 1.8.8 vs. 1.8.7

**Comment by NiLuJe** [comment]
Yeah, the different spacing is the same with the other kerning methods, so, err, possibly because of something else that changed somewhere :D.

**Comment by poire-z** [comment]
Well, with latest crengine master (so, including this PR and not the one that revert it), with HB 1.8.7 and freetype NOT rebuilt against harfuzz, with their 1082.patch and this patch to our base, after cleaning the harfbuzz build tree:
```diff
--- a/thirdparty/harfbuzz/CMakeLists.txt
+++ b/thirdparty/harfbuzz/CMakeLists.txt
@@ -15,10 +15,12 @@ assert_var_defined(FREETYPE_DIR)
 ep_get_source_dir(SOURCE_DIR)
 ep_get_binary_dir(BINARY_DIR)

+set(PATCH_CMD1 sh -c "patch -N -p1 < /tmp/1082.patch || true")
+
 set(CFG_ENV_VAR "CC=\"${CC}\" CXX=\"${CXX}\" CFLAGS=\"${CFLAGS}\" CXXFLAGS=\"${CXXFLAGS}\" LDFLAGS=\"${LDFLAGS}\"")
 set(CFG_ENV_VAR "${CFG_ENV_VAR} FREETYPE_CFLAGS=\"-I${FREETYPE_DIR}/include/freetype2\"")
 set(CFG_ENV_VAR "${CFG_ENV_VAR} FREETYPE_LIBS=\"-L${FREETYPE_DIR} -lfreetype\"")
-set(CFG_OPTS "--prefix=${BINARY_DIR} --enable-shared --disable-static --with-freetype --with-ucdn --without-glib --without-gobject --without-cairo --without-fontconfig --without-icu --without-graphite2 --without-uniscribe --without-directwrite --without-coretext --host=\"${CHOST}\"")
+set(CFG_OPTS "--prefix=${BINARY_DIR} --enable-shared --disable-static --with-freetype --with-ucdn --without-glib --without-gobject --without-cairo --without-fontconfig --without-icu --without-graphite2 --without-uniscribe --without-directwrite --without-coretext --enable-h_advance_cache --host=\"${CHOST}\"")
 set(CFG_CMD sh -c "${CFG_ENV_VAR} ${SOURCE_DIR}/configure ${CFG_OPTS}")

 ko_write_gitclone_script(
@@ -33,7 +35,7 @@ ExternalProject_Add(
     ${PROJECT_NAME}
     DOWNLOAD_COMMAND ${CMAKE_COMMAND} -P ${GIT_CLONE_SCRIPT_FILENAME}
     BUILD_IN_SOURCE 1
-    PATCH_COMMAND NOCONFIGURE=1 ./autogen.sh
+    PATCH_COMMAND COMMAND NOCONFIGURE=1 ./autogen.sh COMMAND ${PATCH_CMD1}
     CONFIGURE_COMMAND ${CFG_CMD}
     BUILD_COMMAND $(MAKE) -j${PARALLEL_JOBS} --silent
     INSTALL_COMMAND $(MAKE) -j${PARALLEL_JOBS} install
```
(not sure about the order of the patch commands...)

**I get back the 13s render time !**
If I just remove `--enable-h_advance_cache`, I get the 55s as before (65s on re-rendering)
No significant increase of memory usage after load, 133M without, 134M with, could be noise.
I haven't checked how the cache gets cleaned or release thus.

So, timing wise, harfbuzz PR 1082 helps a lot!

Now, it's for you to confirm you get the kerning/hinting as you expect/liked it with this PR not reverted (= the same as in your screenshots from 2 days ago).

**Comment by NiLuJe** [comment]
Okay, I'm stupid, the spacing discrepancies were because... I had changed the spacing, and forgot to do that on the new checkout :D.

**Comment by NiLuJe** [comment]
Okay, tried again with matching settings on master, similar results, with a caveat:

Native hinting + harfbuzz shaping is borked when FT is *NOT* built against HB (for... some reason? That doesn't make sense, AFACT, FT+HB should only affect Auto).

(It's okay with FT kerning).

And, of course, an FT+HB build seems to slow things down some more with harfbuzz shaping...

EDIT: Conversely, without FT+HB, it's Auto that's weird, and doesn't match my previous results... -_-".

I'm also not feeling much speedup from the PR ;?

**Comment by NiLuJe** [comment]
TL;DR: Meh?

**Comment by NiLuJe** [comment]
Also, forgot to answer that: yep, wrong patching order: patch first, autreconf/autogen later, especially here, with a patch affecting configure.ac ;)

**Comment by poire-z** [comment]
> wrong patching order: patch first, autreconf/autogen later

OK. It works as well with the right ordering.
(I indeed noticed 2 times the same kind of output while rebuilding - I guess the configure part  noticed configure.ac has changed, and cleverly re-did the autogen part)

**Comment by poire-z** [comment]
Just a note, in case we use 1082.patch, and unrevert the `hb_ft_font_set_load_flags()` calls and call them too in the `setHinting()` method: we'll have to add in that harfbuzz patch `ft_font->advance_cache.clear()` in the `hb_ft_font_set_load_flags()` function too.

**Comment by poire-z** [comment]
@NiLuJe 
Looks like something equivalent to the HB PR 1082 mentionned above has been made into HB 2.0.0:
```
https://github.com/harfbuzz/harfbuzz/pull/1082
https://github.com/harfbuzz/harfbuzz/commit/0f520adaacca3c7b6d8e84a7722343184105f612
https://github.com/harfbuzz/harfbuzz/commit/0ea42e117ba2c76e118974fe114ae5d9ceef5743
https://github.com/harfbuzz/harfbuzz/commit/d8a67dac2a673138bb4d41cd7eab97c9ee987958
https://github.com/harfbuzz/harfbuzz/commit/cbea7d49ab8d4765a2d72dcbf608d326bdf9af3d
https://github.com/harfbuzz/harfbuzz/commit/237f21537842e6b471cdd6c86b98edfc0da0756c
https://github.com/harfbuzz/harfbuzz/commit/047a84c5dd76cdfc072de25c572e30866f87a1f7
https://github.com/harfbuzz/harfbuzz/commit/f90bab8560816b60b4b3f2379b36c08756b21e6c
https://github.com/harfbuzz/harfbuzz/commit/54998befc43ef38e47b74b3153380adbcf6279d4
https://github.com/harfbuzz/harfbuzz/commit/383060cc3354e12611dec3082a6fe08fdb25f652
```

I just had a try with uncommenting the 2` // NOTE: Commented out for now, it's prohibitively expensive. c.f., #230` added in https://github.com/koreader/crengine/pull/231, and my test case from above loads in 13s instead of the 55s from some months ago. So, 13s with or without your FT load flags stuff.

So, you may want to revisit this again :)
(With possibly some more stuff like the one I mentionned at the bottom of https://github.com/koreader/crengine/pull/230#issuecomment-414107254 - I haven't re-read this whole issue :)

**Comment by NiLuJe** [comment]
Yep, I noticed ;).

Haven't had time to check again, though. Plus, IIRC, for best results, we'd have to do the whole "we need FT built against HB" thing.
